### PR TITLE
feat: Complete derive helper attributes

### LINKED ISSUE #18588: Completions for derive macro helper attributes
While rust-analyzer provides completions for derive macros, it doesn't provide completions for the helper attributes they provide. I've tested this both on custom derive macros and the `Default` macro from the standard library.

There seems to be [some code related to this](https://github.com/lowr/rust-analyzer/blob/051c6598bedadcb65710276d1ee853e3743730a3/crates/hir-def/src/nameres/collector.rs#L2060) already.

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
Only their names, anything can go inside.

Closes #18588.

This only completes when you type the first letter, I don't know why...

### CODE REVIEW

File: crates/ide-completion/src/completions/attribute.rs [hunk 1]
```diff
@@ -86,10 +86,26 @@ pub(crate) fn complete_attribute_path(
...
+            SymbolKind::Attribute,
+            ctx.source_range(),
+            format!("{derive_helper}(…)"),
+            ctx.edition,
+        );
+        if let Some(cap) = ctx.config.snippet_cap {
+            item.insert_snippet(cap, format!("{derive_helper}($0)"));
+        } else {
+            item.insert_text(format!("{derive_helper}("));
+        }
```
Comment (Veykril) [inline crates/ide-completion/src/completions/attribute.rs:104]: Derive helpers don't need to accept a token tree, the can be bare `#[helper]` or kv style `#[helper = 0]`, so I don't think we should complete these with `()`

File: crates/ide-completion/src/context/analysis.rs [hunk 2]
```diff
@@ -1129,7 +1129,22 @@ fn classify_name_ref(
...
+                matches!(
+                    kind,
+                    SyntaxKind::STRUCT
+                        | SyntaxKind::ENUM
+                        | SyntaxKind::UNION
+                        | SyntaxKind::VARIANT
+                        | SyntaxKind::TUPLE_FIELD
+                        | SyntaxKind::RECORD_FIELD
+                )
+            })
```
Comment (Veykril) [inline crates/ide-completion/src/context/analysis.rs:1143]: These could be used within the type specifications within ADTs as well, as in `struct T { field: fn(#[helper] ()) }` if I am not mistaken, though this is good enough

**Comment by Veykril** [comment]
> This only completes when you type the first letter, I don't know why...

Empty attributes have a broken parse tree, so thats likely why
`#[]` yields:
```
ATTR@0..3
  POUND@0..1 "#"
  L_BRACK@1..2 "["
  META@2..3
    PATH@2..3
      PATH_SEGMENT@2..3
        ERROR@2..3
          R_BRACK@2..3 "]"
```

File: crates/ide-completion/src/completions/attribute.rs (hunk 1 above)
Comment (ChayimFriedman2) [reply crates/ide-completion/src/completions/attribute.rs:104]: Oh I thought they cannot. I will fix that in a minute.

I really wish proc-macro authors had some way to let r-a analyze their grammar...

File: crates/ide-completion/src/context/analysis.rs (hunk 2 above)
Comment (ChayimFriedman2) [reply crates/ide-completion/src/context/analysis.rs:1143]: I just checked and that's not true.

Comment (Veykril) [reply crates/ide-completion/src/context/analysis.rs:1143]: image
It does work, but no one ever does this. rustfmt even drops the attribute in the function pointer param. So either way this is good enough

**Review by Veykril — APPROVED** [review]
Thanks!

File: crates/ide-completion/src/context/analysis.rs (hunk 2 above)
Comment (ChayimFriedman2) [reply crates/ide-completion/src/context/analysis.rs:1143]: Ah I thought you mean the opposite - that the attribute is not a helper but completed as such.

**Pull request merged**